### ISSUE TITLE
Documentation for running pytest with a different MAIL_BACKEND when using Celery and the app factory pattern

### ISSUE THREAD

**Issue by nickjj** [waynerv/flask-mailman#16]
Hi,

When running tests it's a great idea to use the `locmem` back-end to avoid really sending out emails to your SMTP provider, but if you happen to use Celery and the app factory pattern it's not straight forward on how to accomplish this.

Normally you'd have a pytest fixture like this:

```py
@pytest.fixture(scope="session")
def app():
    params = {
        "DEBUG": False,
        "TESTING": True,
        "WTF_CSRF_ENABLED": False
    }

    _app = create_app(settings_override=params)

    ctx = _app.app_context()
    ctx.push()

    yield _app

    ctx.pop()
```

And if you weren't using Celery you could drop a `MAIL_BACKEND = "locmem"` in the `params` and you're good to go.

But if you're using Celery this changes because when you create your Celery app as part of starting your Flask app, it won't be running with `TESTING = True`, which means it never gets set and suddenly if you have a test that hits a Flask URL that sends an email (reset password, etc.) then the above app fixture never gets used.

Having to set the MAIL_BACKEND in development and restart everything just to run your tests with a different backend doesn't seem like a fun workaround either because you have to remember to keep changing this and often times in dev you want a different backend than test.

How would you solve this problem?

**Comment by rehmanis** [comment]
@nickjj I am not an expert but can't you set the `MAIL_BACKEND` using the [`get_connection()`](https://github.com/waynerv/flask-mailman/blob/master/flask_mailman/__init__.py#L37) function in the test ?

**Comment by nickjj** [comment]
Based on one of your tests using that here: https://github.com/waynerv/flask-mailman/blob/master/tests/test_backend.py#L95-L104 I'm not sure how that could be applied to my tests because these aren't unit tests on sending the mail out.

For example, imagine this flow:

- You set `mail = Mail()` in your Flask app factory and init the extension like usual
- You have a Flask view with a `User.send_reset_password("foo@example.com")` function call
- In your user model you call `current_celery_app.send_task("myapp.user.tasks.deliver_reset_password", ["foo@example.com"])`
- In your task file for that function you call `mail.send(...)` where you send the email with Flask-Mailman

And now you have a test which makes a POST request to reset your password. At this point the mailing is done inside of that task file, not the view so there's no way to override `get_connection()` in the test.

I also wouldn't want to mock out the task function in the test because I'd like to assert the email I'm sending out is has the correct information (correct recipient, subject, template text, etc.).

**Comment by rehmanis** [comment]
In that case pardon me for bring that up. I haven't used Celery ( I just used plain threading module with flask-mailman) so was not aware of this. Might be still not useful suggestion or a bit hacky but can you not use a fixture to yield a new `MAIL_BACKEND` and then after yield restore the `MAIL_BACKEND` to old value ?

**Comment by nickjj** [comment]
Do you have any suggestions on how to set up that fixture?

**Comment by rehmanis** [comment]
I think since MAIL config is already initialized you might have to create  a different app with the config for email. I wrote some test to test my configurations for production and development like this https://github.com/Abdur-rahmaanJ/shopyo/blob/dev/shopyo/tests/conftest.py and this https://github.com/Abdur-rahmaanJ/shopyo/blob/dev/shopyo/tests/test_configs.py which might not be what you want. Is it possible to see your pytest that you wrote?

**Comment by nickjj** [comment]
In the most simple test case, it's:

```python
    def test_home_page(self, app):
        current_celery_app.send_task("myapp.user.tasks.deliver_reset_password", ["foo@examle.com"])
```

This is running in a class that has the original fixture applied from the issue (the "app" one).

As for the Flask factories themselves outside of the tests, here's example `create_app` and `create_celery_app` functions: https://github.com/nickjj/docker-flask-example/blob/505eafcad36c2475d2dd432da665c550d0af33e3/hello/app.py#L14-L60

**Comment by rehmanis** [comment]
I see. So it more of how to setup the celery_app configs and integration. Sorry I was unable to help.

**Comment by nickjj** [comment]
Yeah, but only during tests because it works fine outside of tests (I can control which backend to use with `MAIL_BACKEND`).

**Comment by waynerv** [comment]
I've never used celery, so I'll need some time to investigate the problem you describe.

**Comment by caffeinatedMike** [comment]
Just came across this package while reviewing alternatives to the no-longer-maintained flask-mail. What attracts me to this package is the ability to create custom backends. Locmen is very useful when developing locally. However, in the event of testing with celery I think it would be worthwhile developing a SQLite backend that stores sent emails. This would allow you to check the emails sent across processes (the main/pytest process and the celery worker process)
